subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParallelIntegrationTest.groovy

```diff
@@ -265,13 +285,33 @@ class WorkerExecutorParallelIntegrationTest extends AbstractWorkerExecutorIntegr
         blockingHttpServer.expectConcurrentExecution("workItem1")
 
         expect:
-        args("--max-workers=4")
+        args("--max-workers=3")
         succeeds("parallelWorkTask")
 
         and:
         output.contains("A failure occurred while executing work item 2")
     }
 
+    def "max workers is honored by parallel work"() {
+        given:
+        buildFile << """
+            task parallelWorkTask(type: MultipleWorkItemTask) {
+                doLast {
+                    6.times { i ->
+                        submitWorkItem("workItem\${i}")
+                        sleep 100
+                    }
+                }
+            }
+        """
+        blockingHttpServer.expectConcurrentExecution("workItem0", "workItem1", "workItem2")
```

Comment:
This looks like flakyness waiting to happen, relying on `sleep 100` to order when the actions start. Another option would be to improve the test fixture so that we can assert the behaviour we expect, something like: expect any 3 of these operations to run concurrently, then expect the remaining 3 to run concurrently.

Even better if we could do things like "expect any 3, then unblock 1, then expect another 1 to join, then unblock 2 more, then expect another 2 to join, then unblock them all"